src/gradient.jl

```diff
@@ -28,6 +28,11 @@ end
 
 @inline gradient!(out, f::F, x::SArray) where {F} = vector_mode_gradient!(out, f, x)
 
+@inline function gradient!(out::ImmutableDiffResult, f, x::SArray)
+    dual = vector_mode_dual_eval(f, x)
+    out = DiffResult(value(dual), extract_gradient(dual, x))
+end
```

Comment:
I tried it in a different branch. The performance seems fine. The `!` methods returning a `DiffResult`  have performance within 10% of the non-`!` methods in the (limited) benchmarking I did.

I saw no way to use the existing methods for hessians since they seem to rely on mutating for the value, so I tried this:
```
function hessian!(result::ImmutableDiffResult, f::F, x::SArray) where {F} 
    d1 = dualize(f, x)
    d2 = dualize(f, d1)
    fd2 = f(d2)
    val = value(value(fd2))
    grad = extract_gradient(value(fd2), x)
    hess = extract_jacobian(partials(fd2), x)
    result = DiffBase.hessian!(result, hess)
    result = DiffBase.gradient!(result, grad)
    result = DiffBase.value!(result, val)
end
```
Do you think this method is a workable solution? It has 1 allocation in my tests, not sure why. The performance is fine. For the `prod` example in the tests, it is actually faster than `hessian`.